Nanoemoji build chain with color font outputs for desktop and web

JeppeKlitgaard opened the pull request:
# Introduction
This PR continues the work of @mavit in #260 and fixes the following issues:

Fix: #93 (though this has other discussions in it as well, such as fixing "black fillings")
Fix: #413 
Potentially fixes: #339

# TODO
Before merge:
- [x] Verify fonts work for different platforms and systems (see [comment](https://github.com/hfg-gmuend/openmoji/pull/422#issuecomment-1342808954))
- [x] Fix `font/README.md`
- [x] @b-g Verify that you can also run the new tooling. It should be done the way it usually was done. (`helpers/generate-fonts.sh`)
- [x] Make `SVG+COLR` fonts using `maximum_color`
- [x] Add this information to `README.md`
- [x] Update fonts again
- [x] Fix bug reported by @b-g 

After merge:
- [ ] Make new issue for investigation into ascender/descender sizes
- [ ] Make new issue for investigation into whether `zopflipng` should be disabled
- [x] Make new issue tracking COLRv1 support and deprecation of `cbdt`, `sbix`, etc
- [x] Change over font-builder docker image, needs @b-g or permissions. The Dockerfile is found in `helpers/docker`
- [x] Close #260 

# Content
This PR entirely removes the existing `scfbuild` tooling for rendering fonts. `scfbuild` is old and abandoned whereas the replacement, `nanoemoji`, is actively maintained and implements a lot of features that result in smaller font file sizes and wide compatibility.

We generate fonts for all the formats described below and it is up to the user to choose one that works for their application. A guide will be given in `README.md` in the future.

## Notes
`nanoemoji` does a __lot__ more work under the hood than the `scfbuild` tooling did and it takes substancially longer to run. On my machine, the initial rendering of the fonts takes more than an hour. Luckily `nanoemoji` has a very intelligent build system and renders much quicker for subsequent runs and automagically detects which files have changed and rerenders just those.

## Font formats
`nanoemoji` allows us to render color fonts using a wide variety of formats. A quick overview of them is given here:
- Black
  - `glyf`: Standard format that is widely supported
- Color
  - `cbdt`: Bitmap format, works well for Android and old Chromium-based systems (< 98)
  - `colr_0`: Widely supported color format that works in all modern browsers, but lacks some support for advanced features like gradients
  - `colr_1`: The emerging standard for color fonts, but not yet supported by Safari and many desktop applications
  - `sbix`: Apple-centric format that is supported in Safari and some MacOS apps
  - `svg`: Previously preferred and championed by Apple and Mozilla. Not supported by Chromium-browsers, but works in Safari and Firefox

### Support
For an overview of which applications support which formats, have a look at [colorfonts.wtf](https://www.colorfonts.wtf/), [caniuse/colr](https://caniuse.com/colr), and [caniuse/colr_v1](https://caniuse.com/colr-v1)

# Demo

I've now updated my own site to use the generated OpenMoji fonts. Those who are particularly curious can have a look here: https://jeppe.science/

The emojis have a bit more space around them than the `twemoji` I was using before. This means that the emoji become quite small in regular text and can be somewhat hard to make out, in my opinion. This can possibly be changed by modifying the `ascender` and `descender` sizes?

JeppeKlitgaard commented:
Since the tooling used to generate fonts change from `scfbuild` to `nanoemoji` with this PR, we should verify that the generated fonts work in different platforms

JeppeKlitgaard commented:
I've tested the following on my system (Windows 11):
- Firefox
- Edge
- Brave (should apply to all Chromium-based browsers, including Edge)

Brave and Edge support `glyf`, `cbdt`, `glyf_colr_0`, `glyf_colr_1` and `sbix`, but does NOT support `picosvgz` or `untocuhedsvgz`.
Firefox supports:  `glyf`, `glyf_colr_0`, `glyf_colr_1`, `picosvgz`  and `untocuhedsvgz` but does NOT support `sbix` or `cbdt`.

JeppeKlitgaard commented:
@b-g took a little longer than I had hoped, but I am very pleased with the results.

I'll try adding the font to my personal website and see how that goes. Additionally I'll update `font/README.md` with the new information. 

I believe the only major browser to not support `COLRv1` is Safari, who have recently changed their position on `COLRv1` from negative to neutral. `COLRv0` _should_ work across all sites and OpenMoji does not (afaik) rely on any of the features exclusive to `COLRv1`.

It would be very helpful if you would test it out in Adobe CC (and other places that might be relevant that I don't know about).

b-g commented:
Hi @JeppeKlitgaard, wow!!! This is mega. ❤️

Comments and questions:

- I checked the fonts on Safari / Firefox and added my marks to the table ... looks good on macOS
- reports for Adobe CC still pending
- the config files `OpenMoji-Color.ttx` `scfbuild-color.yml` are missing, couldn't find them. How can I set meta data things like name, version etc for the fonts?
- shellcheck reports warnings in the bash scripts, please check and review
- What is the github action doing exactly?
- running the build system via docker works for me. Yay! But it took ca. 7h, it this normal? Could it be speeded up by running things in paralell?
- And last could I get your "promise" to help in case thinks break unforeseeable due to the new cool font build system? :) 

Happy weekend!

JeppeKlitgaard commented:
Thank you very much. A lot of the groundwork was done by @mavit – he deserves a lot of praise!

> the config files OpenMoji-Color.ttx scfbuild-color.yml are missing, couldn't find them. How can I set meta data things like name, version etc for the fonts?

`scfbuild-color.yml` has been removed entirely since it is no longer used – this PR removes `scfbuild` entirely. @mavit's original work used both, but I think `nanoemoji` is more mature now and can be relied upon, while `scfbuild` has only gotten older. The additional cognitive overhead and maintenance burden of maintaining two build chains is doable, but something I would caution against.

I have moved `OpenMoji-Color.ttx` into the `data` folder, but otherwise not changed anything. I think it fits better with the content of that folder and ensures that the `font` folder doesn't have any files that could confuse someone looking to simply use the generated fonts. 

The `font` folder now looks fairly neat and when the `README.md` file is updated should be easy to get an overview of even for people that aren't experienced with GitHub:
![image](https://user-images.githubusercontent.com/1814611/206849707-7e8409d1-1129-4a12-b15c-f0297e9f2f28.png)

> shellcheck reports warnings in the bash scripts, please check and review

Well spotted! I have fixed both warnings now.

> What is the github action doing exactly?

The GitHub action builds the container that we use for building the fonts. It only runs if the Dockerfile has changed. We would need to set up a new hosted image owned by the `hfg-gmuend` organisation such that any updates to the Dockerfile would automatically make a new image available. I'm happy to help set this up in a way that should lead to a very easy 'set it and forget it' future usage.

> running the build system via docker works for me. Yay! But it took ca. 7h, it this normal? Could it be speeded up by running things in paralell?

7 hours seems long – are you running it on an M1 Mac by chance? That would lead to a performance slow-down due to the Docker image being x86, thus requiring emulation.

It should be noted that `nanoemoji` does _a lot_ more things than `scfbuild` did, many of which we probably want:
- Many more formats are generated (glyf, cbdt, glyf_colr_0, glyf_colr_1, sbix, svg) as opposed _just_ svg.
- `nanoemoji` looks for shapes that are present in multiple icons and reuses them, which leads to smaller font sizes. It also implements other tricks to make the SVG's as small as possible (this leads to a file size reduction from 4.4MB → 2.9MB!)
- `nanoemoji` generates bitmap images for the non-SVG formats (cbdt, sbix). These bitmaps are compressed using `pngquant` and `zopflipng` (24MB → 17MB → 17MB)
- `nanoemoji` generates `colr_0` and `colr_1` formats from the SVGs
- The script additionally does `woff2` compression to make fonts ready for use in web browsers (≈ 25% file reduction, if I recall correctly)

Any subsequent runs you do should be a lot faster since `nanoemoji` is clever enough to reuse the files generated in all these intermediate steps and only does the final font 'assembly' again.

Ways to reduce build times could include: 
- Disable `zopflipng` – it doesn't seem to lead to a file size reduction anyway at the moment
- Reduce the number of formats generated
  - Most people will be using `colr_0` and `svg` for now, which gives a lot of compatibility.
  - I do think it would be nice to generate all the formats. Particularly desktop applications are fuzzy about which formats they use and they generally have lower adoption of both `colr_0` and `colr_1` formats and instead rely on the bitmap formats.
- Check the build files into git to allow reuse of the inter

…
NOTE: I have to run, but leaving this here for now. Will update comment when back

mavit commented:
Platform – System | Tester – Comment | black-glyf | color-cbdt | color-glyf_colr_0 | color-glyf_colr_1 | color-picosvgz | color-sbix | color-untouchedsvgz
-- | -- | -- | -- | -- | -- | -- | -- | --
Chromium 108 – Fedora 37 | @mavit | ✔ | ✔ | ✔ | ✔ | x | ✔ | x
Firefox 108 - Fedora 37  | @mavit | ✔ | x | ✔ | ✔ | ✔ | x | ✔
Gnome 43.2 - Fedora 37   | @mavit | ✔ | ✔ | ✔ | [x](https://gitlab.freedesktop.org/cairo/cairo/-/merge_requests/370) | [x](https://gitlab.freedesktop.org/cairo/cairo/-/merge_requests/319) | ✔ | [x](https://gitlab.freedesktop.org/cairo/cairo/-/merge_requests/319)

mavit commented:
So far, everything that’s been tested recently supports COLRv0.  However, if we discover an application that doesn’t support COLRv0, something to consider is `nanoemoji`’s new [`maximum_color` feature](https://github.com/googlefonts/nanoemoji/blob/main/README.md#adding-color-tables-to-existing-fonts), which bundles multiple formats into a single `.ttf`.  This increases the size, of course, but I think that’s only a major concern for web fonts.

b-g commented:
With the Feedback on the fonts in Adobe CC by Daniel (see https://github.com/hfg-gmuend/openmoji/pull/260#issuecomment-1362525109) ... I think we are all set here. IMO the new build system and fonts definitely outperform what we currently have! Yay!! Many thanks @mavit and @JeppeKlitgaard for this achievement! 🙏 ❤️

Happy to merge this PR once I get to go of @JeppeKlitgaard. Any last improvements before we merge?

+ Merry 🎄 + 🛷 2023!

b-g commented:
@JeppeKlitgaard Ping :) Hope you've had a nice start into 2023! 🛷

> Happy to merge this PR once I get to go of @JeppeKlitgaard. Any last improvements before we merge?

JeppeKlitgaard commented:
> @JeppeKlitgaard Ping :) Hope you've had a nice start into 2023! 🛷
> 
> > Happy to merge this PR once I get to go of @JeppeKlitgaard. Any last improvements before we merge?

I hope you managed to enter the New Year in a pleasant fashion as well! I have been meaning to finish this up, but admittedly forgot about it slightly. I will have a look later today or (worst case) over the weekend. I think there are a few final touches that we should do

JeppeKlitgaard commented:
While `maximum_color` is still experimental and not particularly well-documented, I think it might make sense to add it.

Additionally, we should set up the container image to be automatically built and pushed, though this is probably more manageable to do after this is merged

JeppeKlitgaard commented:
I have rebased off of `master`, but still need to run and commit new `SVG+COLR` fonts made using `maximum_color` as well as update the other fonts with changes from `master`.

JeppeKlitgaard commented:
I am running the builder now, though perhaps @b-g could rerun it on his machine to confirm everything is working once I've pushed the last commit and removed the WIP marker from the PR (likely tonight).

waldyrious commented on helpers/generate-index-html.js:
```diff
@@ -202,12 +202,12 @@ li.selected {
-    src: url('font/OpenMoji-Black.ttf') format('truetype');
-}
+    src: url('font/glyf/OpenMoji-Black.ttf') format('truetype');
+  }
```
Why adding two spaces to this line? I'd propose keeping it as-is:

```suggestion
}
```

Also, heads-up: whitespace in this file is changed in #428.

JeppeKlitgaard replied:
You would simply need to rebase off of this in your PR for the changes to whitespace.

I agree that the whitespace is superfluous and will update the final commit to remove this. Thanks for reviewing!

waldyrious commented on helpers/generate-index-html.js:
```diff
@@ -202,12 +202,12 @@ li.selected {
-    src: url('font/OpenMoji-Color.ttf') format('truetype');
-}
+    src: url('font/glyf_colr_0/OpenMoji-Color.ttf') format('truetype');
+  }
```
(Same as above.)

JeppeKlitgaard commented:
@b-g this is ready to merge now. You may want to try and run the builder on your end again, though nothing substantial has changed and it is working nicely for me. The biggest issue is definitely the build times, though with cache it isn't __terrible__ and it only needs to be done when making new releases.

waldyrious commented on index.html:
```diff
@@ -192,12 +192,12 @@
-    src: url('font/OpenMoji-Black.ttf') format('truetype');
-}
+    src: url('font/OpenMoji-black-glyf/OpenMoji-black-glyf.woff2') format('woff2');
+  }
```
Tiny nit: you forgot to regenerate `index.html` after the change in the generator script :)

JeppeKlitgaard replied:
Done

JeppeKlitgaard commented:
It would be helpful to merge this before too many more changes, as rebasing can be a bit laborious. Should ideally be merged ahead of #431, for example.

b-g commented:
@JeppeKlitgaard Many thanks! Near there! Yay! 

Comments:
- Shellcheck reports issues in `helpers/generate-fonts-runner.sh`, please fix.
- I had quite a few merge conflicts, tried to resolve them. Should be good now ... but please double check.
- Currently running `npm run generate-font` ... looks good so far. Will report once the build run is finished.

b-g commented:
- `npm run generate-font` finished. Yay! (Runtime ca. 5-6h on my mac)

JeppeKlitgaard commented:
I've fixed the shellcheck warnings. I assume running `npm run generate-font` resulted in new font files – could you pull my changes for the shellcheck warnings and then commit the new fonts and push that? Alternatively I am happy to run the builder again tomorrow afternoon and do the commit the final font files after which we should be ready to merge this PR.

I thought I had rebased off of master and fixed any merge conflicts, but I must not have done that. I've had a look at the merge resolution you did and it all seems fine to me – I manually had a peek at most of the SVGs and the other file changes seem to be correct as well. Apologise for the inconvenience there!

b-g commented:
Hi @JeppeKlitgaard, 

> npm run generate-font resulted in new font files

Yes, but I was slight confused as the generated files only appeared in the `/build/fonts` but were not copied automatically to the font folder.

Also there very last line of the entire build process seems to report an issue `rm: cannot remove '/mnt/build/fonts/OpenMoji-color-colr0_svg': No such file or directory`. Please find attached my entire log: [Terminal Saved Output.txt](https://github.com/hfg-gmuend/openmoji/files/10380302/Terminal.Saved.Output.txt)

Hence what about the following approach:
We merge the current state of the PR, even though still a bit WIP ... it seems to be easier to iterate with the changes in in place in the master branch e.g. to put the docker publishing in place. Just it would be good to finish up quickly to not leave master in an in-between state. OK / do you have time in the next days?

JeppeKlitgaard commented:
@b-g This is definitely my fault. This bug was introduced with the multi-format font support I made over the weekend and since the build is so a fairly lengthy affair especially without cache I never did a full-rebuild which would have revealed this bug. I will push for a fix for this before the merge. It would probably be good if I could do a full rebuild myself (i.e. deleting `build` and running `npm run generate-font`) before the merge as well. If I remember I will do this overnight.

JeppeKlitgaard commented:
@b-g I have now done a full rebuild from scratch (i.e. deleting `./build` and the fonts in `./font` and everything seems to be working.

This should be mature enough to merge now

b-g commented:
@JeppeKlitgaard gave it a go as well! Runs now smooth! Many many thanks, also to @mavit! Merged! Yay!! 🎉🎉🎉!
